src/parallel/parallel.hpp

```diff
@@ -53,37 +54,40 @@ class SegmentLock {
     // thread on the lock. However, if the number of locks exceeds an
     // unspecified limit, then it will throw a std::system_error.
     std::recursive_mutex mutex_;
-    // N.B. Using counter_ = 0 causes the linter to complain that the copying
+    // N.B. Using version_ = 0 causes the linter to complain that the copying
     //      invokes a deleted constructor
-    // N.B. Using counter_(0) causes the linter to complain, thinking counter_
-    //      is a function.
-    AtomicCounter counter_{0};
+    // N.B. Using version_(0) will cause the compile to parse version_ as a
+    //      function.
+    // N.B. This needs to be atomic since it can be read from multiple threads
+    //      via get_version() while its being modified by one thread.
+    AtomicVersion version_{0};
+    unsigned locked_count_{0};
 };
 
 ////////////////////////////////////////////////////////////////////////////////
 /// HELPER CLASSES
 ////////////////////////////////////////////////////////////////////////////////
 
+static constexpr KeyType bucket_empty_key = static_cast<KeyType>(-1);
+static constexpr KeyType bucket_locked_key = static_cast<KeyType>(-2);
+
 /// We are assuming this will be packed into 8 (or 16... if we've changed the
 /// size of the KeyType/ValueType to int64_t) bytes so that we can atomically
 /// update both.
 struct KeyValue {
-    KeyType key = 0;
+
+    KeyType key = bucket_empty_key;
     ValueType value = 0;
+
+    constexpr bool operator==(const KeyValue &) const = default;
```

Comment:
Non-blocking but what does this do? I've never understood all this funky syntax

Reply:
Basically just asks the compiler to generate an `==` operator using the "default" common-sense understanding of value equality.